### PR TITLE
calendar#523 added event object to ICS translator

### LINKED ISSUE #523: [TASK] add event to ICS translator
in the dav-parser add a `translate` method that translates a calendar event object to an ICS

### PR DESCRIPTION
**Pull request by rezk2ll**
closes https://github.com/OpenPaaS-Suite/esn-frontend-calendar/issues/523

### CODE REVIEW

File: src/lib/translator.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
+import { CalendarEventObject, Attendee, Dictionary, RRule } from '../types/dav-parser';
+import { Component, Property } from 'ical.js';
+
+/**
+ * Translates a CalendarEventObject into an ICS String.
+ * 
+ * @param event CalendarEventObject - the event object to convert to ICS
```
Comment (itswillta) [inline src/lib/translator.ts:7]: I think you are using too many different terms here. Sometimes you say "translate A into B", the other times you say "convert A to B", "transform A into B", "build B". In my opinion, you should just use 1 term.

File: src/lib/translator.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  calendarComponent.addPropertyWithValue('calscale', 'GREGORIAN');
+  // add the event component as a subcomponent for the calendar component
+  calendarComponent.addSubcomponent(eventComponent);
+
+  // transform to ICS
+  return calendarComponent.toString();
+}
+
+/**
+ * transforms an event object into a ICAL component
```
Comment (itswillta) [inline src/lib/translator.ts]: The first letter needs to be uppercased.

File: src/lib/translator.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ */
+const buildEventComponent = (event: CalendarEventObject): Component => {
+  const eventComponent = new Component('vevent');
+
+  eventComponent.addPropertyWithValue('uid', event.id);
+  eventComponent.addPropertyWithValue('summary', event.title);
+  eventComponent.addPropertyWithValue('location', event.location || '');
+  eventComponent.addPropertyWithValue('description', event.description || '');
+  eventComponent.addPropertyWithValue('dtstart', event.start as string);
+  eventComponent.addPropertyWithValue('dtend', event.end as string);
```
Comment (itswillta) [inline src/lib/translator.ts]: Why are you assuming `event.start` and `event.date` will always be `string`s? They can also be `Date`s, can't they?

File: src/lib/translator.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    eventComponent.addProperty(translateAttendee(attendee));
+  });
+
+  // Handle the RRule transformation
+  if (event.rrule) {
+    eventComponent.addProperty(translateRRule(event.rrule));
+  }
+
+  // Handle the alarm translation
+  if (event.alarm && event.alarm.trigger) {
```
Comment (itswillta) [inline src/lib/translator.ts]:
```suggestion
  if (event.alarm?.trigger) {
```

File: test/const.ts [hunk 5]
```diff
@@ -130,3 +132,18 @@ SEQUENCE:0
 END:VEVENT
 END:VCALENDAR
 `;
+
+
+export const testEventObject: CalendarEventObject = {
+  id: '123456',
+  title: 'test',
+  start: '2021-03-15T10:00:00',
+  end: '2021-03-15T11:00:00',
```
Comment (itswillta) [inline test/const.ts:141]: I think you should add a case where `start` and `end` are `Date`s.

File: src/lib/translator.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ * @param event CalendarEventObject - the event object to convert to ICS
+ * @returns String  -  the ICS string representing the event
+ */
+export function translate(event: CalendarEventObject): string {
+  const calendarComponent: Component = new Component('vcalendar');
+  const eventComponent: Component = buildEventComponent(event);
+
+  calendarComponent.addPropertyWithValue('version', '2.0');
+  calendarComponent.addPropertyWithValue('calscale', 'GREGORIAN');
+  // add the event component as a subcomponent for the calendar component
```
Comment (renaudboyer) [inline src/lib/translator.ts]: In my opinion, the code is self-explanatory, no need to add so many comment inside the functions

File: test/translator.test.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+SUMMARY:Daily Meeting - Team Calendar/Contacts\r\n\
+DESCRIPTION:This is an automatic alarm sent by OpenPaas\r\n\
+END:VALARM\r\n\
+END:VEVENT\r\n\
+END:VCALENDAR';
+
+    expect(translate(eventObjectWithAlarm)).toEqual(ics);
+  });
+
+  it('should translate an event object with RRule properly', () => {
```
Comment (renaudboyer) [inline test/translator.test.ts:123]: We also need multi VEVENT event tests

Comment (rezk2ll) [reply test/translator.test.ts:123]: we can't have multi vevent if we are translating one event object if i'm not wrong

edit: exceptions

File: src/lib/translator.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+};
+
+/**
+ * Translates a CalendarEventObject attendee to a JCAL property
+ *
+ * @param attendee Attendee
+ * @returns ICAL.Property
+ */
+const translateAttendee = (attendee: Attendee): Property => {
+  const { email, partstat, rsvp, role, cutype, cn } = attendee;
```
Comment (renaudboyer) [inline src/lib/translator.ts]: We should not specify parameters by name as it's possible to use custom parameters.

So we should pick every parameters, the only exception is email which is the ATTENDEE property value.

**Review by itswillta — APPROVED** [review]

**Review by asmaNode — APPROVED** [review]

**Pull request merged**